### PR TITLE
Adding support for dict in  infer_shape

### PR DESCRIPTION
**Pull request by spidyDev**
Currently in MXNET if I have below code, infer shape fails:

```
data = mx.sym.Variable('data-gpu0')
prev = mx.sym.Variable('prev')
fc1  = mx.sym.FullyConnected(data=data, name='fc1', num_hidden=128)
fc2  = mx.sym.FullyConnected(data=prev, name='fc2', num_hidden=128)
out  = mx.sym.Activation(data=mx.sym.elemwise_add(fc1, fc2), act_type='relu')
out.infer_shape(data-gpu0=(10,64), prev=(10,128))

  File "<ipython-input-6-3834d8e9b880>", line 2
    out.infer_shape(data-gpu0=(10,64), 'prev'=(10,128))
SyntaxError: keyword can't be an expression
```
Also one cannot infer_shape using variable  as name  i.e
```
data = mx.sym.Variable('data')
prev = mx.sym.Variable('prev')
fc1  = mx.sym.FullyConnected(data=data, name='fc1', num_hidden=128)
fc2  = mx.sym.FullyConnected(data=prev, name='fc2', num_hidden=128)
out  = mx.sym.Activation(data=mx.sym.elemwise_add(fc1, fc2), act_type='relu')
name="data"
out.infer_shape(name=(10,64), prev=(10,128))

MXNetError: [10:24:51] src/c_api/c_api_symbolic.cc:422: InferShapeKeyword argument name name not found.
Candidate arguments:
	[0]data
	[1]fc1_weight
	[2]fc1_bias
	[3]prev
	[4]fc2_weight
	[5]fc2_bias
```

To solve this issue I propose solution in this PR to add support for infer_shape to accept dict, so that user can provide data_names and shapes freely.(This will also be similar to mx.viz.plot_network that takes inputs as dict)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add infer_shape to accept "dict" of input names and shapes as key value pair.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@sandeep-krishnamurthy  @lupesko  @nswamy

### CODE REVIEW

**Comment by piiswrong** [comment]
`out.infer_shape(data-gpu0=(10,64), prev=(10,128))`
This is simply a syntax error. You can't use `-` in variable names. You need to use `_`

File: python/mxnet/symbol/symbol.py [hunk 1]
```diff
@@ -1078,11 +1078,22 @@ def _infer_shape_impl(self, partial, *args, **kwargs):
...
-                    if not isinstance(s, tuple):
-                        raise TypeError("Arguments need to be shapes (tuple), "
-                                        "but argument %d is %s." % (i, type(s)))
-                    sdata.extend(s)
-                indptr.append(len(sdata))
+                    if isinstance(s, dict):
+                        str_keys = []
+                        for k, v in s.items():
+                            if not isinstance(v, tuple):
+                                raise TypeError("Arguments need to be shapes (tuple), "
```
Comment (sandeep-krishnamurthy) [inline python/mxnet/symbol/symbol.py]: Thanks for this fix.

1. Same checks in 2 places. Please refactor.
2. Validation on k?
3. indptr.append(..) can be moved of this logic.

**Comment by szha** [comment]
Sounds like we should instead limit the string being used in symbol's variable name.

**Comment by spidyDev** [comment]
@piiswrong  But that is variable name, which is a string . 
mx.sym.Variable('data-gpu0')
User should be able to use special characters. 
Moreover, we have encountered models (ONNX models) that have input name as "data/gpu0", which when imported to MXNet and tried to infer_shape  throws same error.

**Comment by szha** [comment]
This is a conflict between the requirement of variable names in python and the one in our graph representation. Given the use case, I think it's indeed necessary to support dictionary to break the coupling.

**Comment by piiswrong** [comment]
You can use **kwargs for this. No need to change the interface

**Comment by spidyDev** [comment]
@piiswrong  Thanks. That should do the trick. I agree no need to change the interface.

**Pull request closed without merging**